src/divide_conquer_functions/conquer.jl

```diff
@@ -0,0 +1,232 @@
+		predicates,
+		grammar,
+		symboltable,
+		false,
+	)
+	features = float.(features)
+	# Take labels and features to make DecisionTree
+	# See decision tree example: https://github.com/Herb-AI/HerbSearch.jl/blob/subset-search/src/subset_iterator.jl
+	model = DecisionTreeClassifier()
+	fit!(model, features, labels)
+	final_program = construct_final_program(model.root.node, idx_ifelse, solutions, predicates)
+	return final_program
+end
+
+"""
+	Returns predicates that can serve as conditional statements for combining programs in a decision tree.
+
+	# Arguments
+	- `n_predicates`: Determines the number of predicates that are returned.
+	- `sym_bool`: is used to make sure that only programs evaluate to `Bool` are considered. 
+	- `sym_constraint`: Further constrains the program space to exclude trivial predicates.
+"""
+function get_predicates(grammar::AbstractGrammar,
+	sym_bool::Symbol,
+	sym_constraint::Symbol,
+	n_predicates::Number,
+)::Vector{RuleNode}
+	# We get the first grammar rule that has the specified `sym_constraint` and add constraint to grammar. 
+	clearconstraints!(grammar)
+	# Create DomainRuleNode that contains all rules of type sym_constraint and add constraint to grammar
+	rules = grammar.bytype[sym_constraint]
+	domain = DomainRuleNode(grammar, rules)
+	addconstraint!(grammar, ContainsSubtree(domain))
+
+	iterator = BFSIterator(grammar, sym_bool)
+	predicates = Vector{RuleNode}()
+
+	for (i, candidate_program) ∈ enumerate(iterator)
+		candidate_program = freeze_state(candidate_program)
+		push!(predicates, candidate_program)
+		if i >= n_predicates
+			break
+		end
+	end
+	return predicates
+end
+
+"""
+	Returns predicates that can serve as conditional statements for combining programs in a decision tree.
+
+	# Arguments
+	- `n_predicates`: Determines the number of predicates that are returned.
+	- `sym_bool`: is used to make sure that only programs evaluate to `Bool` are considered. 
+	- `rules`: List of grammer rules used to further constrain the program space to exclude trivial predicates.
+"""
+function get_predicates(grammar::AbstractGrammar,
+	sym_bool::Symbol,
+	rules::Vector{Int},
+	n_predicates::Number,
+)::Vector{RuleNode}
+	clearconstraints!(grammar)
+	# Create DomainRuleNode that contains all rules and add constraint to grammar
+	domain = DomainRuleNode(grammar, rules)
+	addconstraint!(grammar, ContainsSubtree(domain))
+
+
+	iterator = BFSIterator(grammar, sym_bool)
+	predicates = Vector{RuleNode}()
+
+	for (i, candidate_program) ∈ enumerate(iterator)
+		candidate_program = freeze_state(candidate_program)
+		push!(predicates, candidate_program)
+		if i >= n_predicates
+			break
+		end
+	end
+	return predicates
+end
+
+"""
+	Returns a matrix containing the feature vectors for all problem/predicate combinations. 
+	A feature vector is obtained by evaluating a `IOExample` in `ioexamples_solutions` on each
+	predicate.
+"""
+function get_features(
+	ioexamples_solutions::Vector{Tuple{T, Vector{Int}}},
```

Comment:
Do we need the solutions too here? Might be cleaner to pass `AbstractVector{<:IOExample}` directly.

Reply:
We need to convert the `problems_to_solutions` map to vectors. Instead of one vector`Vector{Tuple{T, Vector{Int}}}` holding both input-output examples and solutions, I can make a vector with all input-output examples `Vector{<:IOExample}` and a vector with all the solutions. That's easy. My only question here is what benefit `AbstractVector` adds in this case? Do we need this flexibility? What do you think @ReubenJ ?

Reply:
Makes sense after the discussion in the meeting.